File: web-sdk/src/data/services/yalo-message/yalo-message-service-websocket.test.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { Err, Ok, type Result } from '@domain/common/result';
import {
  MessageRole,
  MessageStatus,
  PollMessageItem,
  SdkMessage,
} from '@domain/models/events/external_channel/in_app/sdk/sdk_message';
import type { TokenRepository } from '@data/repositories/token/token-repository';
import { YaloMessageServiceWebSocket } from './yalo-message-service-websocket';

const sockets: MockSocket[] = [];

class MockSocket extends EventTarget {
  readyState = 0;
  url: string;
  send = vi.fn();
  close = vi.fn(() => {
    this.readyState = 3;
    this.dispatchEvent(new Event('close'));
  });
  constructor(url: string) {
    super();
    this.url = url;
    sockets.push(this);
  }
}

const stubWebSocket = () => {
  vi.stubGlobal(
    'WebSocket',
    Object.assign(MockSocket, { CONNECTING: 0, OPEN: 1, CLOSING: 2, CLOSED: 3 })
  );
};

const open = (s: MockSocket) => {
  s.readyState = 1;
  s.dispatchEvent(new Event('open'));
};
const closed = (s: MockSocket) => {
  s.readyState = 3;
  s.dispatchEvent(new Event('close'));
};
const message = (s: MockSocket, data: string) =>
  s.dispatchEvent(new MessageEvent('message', { data }));

const makeTokenRepository = (
  result: Result<string> = new Ok('jwt.token.value')
): TokenRepository => ({
  getToken: vi.fn().mockResolvedValue(result),
});

const makeTextMessage = (text: string): SdkMessage => ({
  correlationId: `cid-${text}`,
  textMessageRequest: {
    content: {
      timestamp: new Date('2026-01-01T00:00:00Z'),
      text,
      status: MessageStatus.MESSAGE_STATUS_IN_PROGRESS,
      role: MessageRole.MESSAGE_ROLE_USER,
    },
    timestamp: new Date('2026-01-01T00:00:00Z'),
  },
  timestamp: new Date('2026-01-01T00:00:00Z'),
});

const flush = () => vi.advanceTimersByTimeAsync(0);

describe('YaloMessageServiceWebSocket', () => {
  const baseUrl = 'api.example.com';
  const wsUrl = `wss://${baseUrl}/websocket/v1/connect/webchat`;

  beforeEach(() => {
    vi.useFakeTimers();
    sockets.length = 0;
    stubWebSocket();
  });

  afterEach(() => {
    vi.restoreAllMocks();
    vi.useRealTimers();
  });

  describe('subscribe', () => {
    it('opens a WebSocket with the token as a query param', async () => {
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());

      service.subscribe(() => {});
      await flush();

      expect(sockets).toHaveLength(1);
      expect(sockets[0].url).toBe(`${wsUrl}?token=jwt.token.value`);
    });

    it('URL-encodes the token', async () => {
      const service = new YaloMessageServiceWebSocket(
        baseUrl,
        makeTokenRepository(new Ok('a/b+c=d'))
      );

      service.subscribe(() => {});
      await flush();

      expect(sockets[0].url).toBe(`${wsUrl}?token=a%2Fb%2Bc%3Dd`);
    });

    it('invokes the callback with parsed PollMessageItem on incoming frames', async () => {
      const callback = vi.fn();
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());

      service.subscribe(callback);
      await flush();
      open(sockets[0]);

      const frame = JSON.stringify(
        PollMessageItem.toJSON({
          id: 'msg-1',
          userId: 'user-1',
          status: 'delivered',
          date: new Date('2026-01-01T00:00:00Z'),
          message: makeTextMessage('hello'),
        })
      );
      message(sockets[0], frame);

      expect(callback).toHaveBeenCalledTimes(1);
      expect(callback.mock.calls[0][0]).toMatchObject({
        id: 'msg-1',
        userId: 'user-1',
        status: 'delivered',
      });
    });

    it('ignores malformed frames without throwing', async () => {
      const callback = vi.fn();
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());

      service.subscribe(callback);
      await flush();
      open(sockets[0]);
      message(sockets[0], 'not-json');

      expect(callback).not.toHaveBeenCalled();
    });

    it('schedules a reconnect when the token fetch fails', async () => {
      const tokenRepository = makeTokenRepository(new Err(new Error('auth failed')));
      const service = new YaloMessageServiceWebSocket(baseUrl, tokenRepository);

      service.subscribe(() => {});
      await flush();
      expect(sockets).toHaveLength(0);

      await vi.advanceTimersByTimeAsync(1000);
      expect(tokenRepository.getToken).toHaveBeenCalledTimes(2);
    });
  });

  describe('sendMessage', () => {
    it('sends the frame as SdkMessage JSON when the socket is open', async () => {
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());
      service.subscribe(() => {});
      await flush();
      open(sockets[0]);

      const result = await service.sendMessage(makeTextMessage('hi'));

      expect(result.ok).toBe(true);
      expect(sockets[0].send).toHaveBeenCalledTimes(1);
      expect(JSON.parse(sockets[0].send.mock.calls[0][0])).toMatchObject({
        correlationId: 'cid-hi',
        textMessageRequest: { content: { text: 'hi' } },
      });
    });

    it('queues frames sent before the socket opens, then flushes on open', async () => {
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());
      service.subscribe(() => {});
      await flush();

      const r1 = await service.sendMessage(makeTextMessage('one'));
      const r2 = await service.sendMessage(makeTextMessage('two'));

      expect(r1.ok).toBe(true);
      expect(r2.ok).toBe(true);
      expect(sockets[0].send).not.toHaveBeenCalled();

      open(sockets[0]);

      expect(sockets[0].send).toHaveBeenCalledTimes(2);
    });

    it('connects on first send when not yet subscribed', async () => {
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());

      const result = await service.sendMessage(makeTextMessage('hi'));
      await flush();

      expect(result.ok).toBe(true);
      expect(sockets).toHaveLength(1);

      open(sockets[0]);
      expect(sockets[0].send).toHaveBeenCalledTimes(1);
    });
  });

  describe('reconnect', () => {
    it('reconnects after the socket closes, with exponential backoff', async () => {
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());
      service.subscribe(() => {});
      await flush();
      open(sockets[0]);

      closed(sockets[0]);
      expect(sockets).toHaveLength(1);

      await vi.advanceTimersByTimeAsync(1000);
      expect(sockets).toHaveLength(2);

      closed(sockets[1]);
      await vi.advanceTimersByTimeAsync(1500);
      expect(sockets).toHaveLength(2);

      await vi.advanceTimersByTimeAsync(500);
      expect(sockets).toHaveLength(3);
    });

    it('caps the backoff delay at 30 seconds', async () => {
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());
      service.subscribe(() => {});
      await flush();

      for (let i = 0; i < 8; i++) {
        const delay = Math.min(30000, 1000 * 2 ** i);
        closed(sockets[i]);
        await vi.advanceTimersByTimeAsync(delay);
      }

      closed(sockets[8]);
      await vi.advanceTimersByTimeAsync(30000);
      expect(sockets).toHaveLength(10);
    });

    it('resets the backoff after a successful open', async () => {
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());
      service.subscribe(() => {});
      await flush();

      closed(sockets[0]);
      await vi.advanceTimersByTimeAsync(1000);
      closed(sockets[1]);
      await vi.advanceTimersByTimeAsync(2000);
      open(sockets[2]);
      closed(sockets[2]);

      await vi.advanceTimersByTimeAsync(1000);
      expect(sockets).toHaveLength(4);
    });
  });

  describe('unsubscribe', () => {
    it('closes the socket and stops reconnect attempts', async () => {
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());
      service.subscribe(() => {});
      await flush();
      open(sockets[0]);

      service.unsubscribe();
      expect(sockets[0].close).toHaveBeenCalled();

      await vi.advanceTimersByTimeAsync(60000);
      expect(sockets).toHaveLength(1);
    });

    it('clears queued frames so they are not sent on a later subscribe', async () => {
      const service = new YaloMessageServiceWebSocket(baseUrl, makeTokenRepository());
      service.subscribe(() => {});
      await flush();

      await service.sendMessage(makeTextMessage('queued'));
      service.unsubscribe();

      service.subscribe(() => {});
      await flush();
      open(sockets[1]);

      expect(sockets[1].send).not.toHaveBeenCalled();
    });
  });
});

File: flutter-sdk/test/data/services/yalo_message/yalo_message_service_websocket_test.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'dart:async';
import 'dart:convert';

import 'package:fake_async/fake_async.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mocktail/mocktail.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:yalo_chat_flutter_sdk/src/common/result.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_message/yalo_message_service_websocket.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_message_auth/token_entry.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_message_auth/yalo_message_auth_service.dart';
import 'package:yalo_chat_flutter_sdk/src/domain/models/events/external_channel/in_app/sdk/sdk_message.pb.dart';

class MockYaloMessageAuthService extends Mock
    implements YaloMessageAuthService {}

class _FakeSink implements WebSocketSink {
  final List<dynamic> sent = [];
  bool closed = false;

  @override
  void add(dynamic data) => sent.add(data);

  @override
  Future<void> close([int? closeCode, String? closeReason]) async {
    closed = true;
  }

  @override
  void addError(Object error, [StackTrace? stackTrace]) {}

  @override
  Future<dynamic> addStream(Stream stream) async {}

  @override
  Future<dynamic> get done => Future.value();
}

class _FakeChannel implements WebSocketChannel {
  _FakeChannel(this.uri);

  final Uri uri;
  final StreamController<dynamic> _incoming = StreamController<dynamic>();
  final _FakeSink _sink = _FakeSink();

  void emit(String frame) => _incoming.add(frame);
  void emitError(Object error) => _incoming.addError(error);
  Future<void> closeStream() => _incoming.close();

  @override
  WebSocketSink get sink => _sink;

  @override
  Stream<dynamic> get stream => _incoming.stream;

  List<dynamic> get sent => _sink.sent;
  bool get sinkClosed => _sink.closed;

  @override
  dynamic noSuchMethod(Invocation invocation) => super.noSuchMethod(invocation);
}

TokenEntry _tokenEntry(String token) => TokenEntry(
  accessToken: token,
  refreshToken: 'refresh',
  expiresAt: DateTime.now().add(const Duration(hours: 1)),
  userId: 'user-1',
);

SdkMessage _textMessage(String text) => SdkMessage(
  correlationId: 'cid-1',
  textMessageRequest: TextMessageRequest(
    content: TextMessage(
      text: text,
      role: MessageRole.MESSAGE_ROLE_USER,
      status: MessageStatus.MESSAGE_STATUS_SENT,
    ),
  ),
);

const _incomingFrame = '''
{
  "id": "3a73bc83-b4da-4fef-a83e-21310edc0283",
  "message": {
    "timestamp": "2026-03-26T16:51:35.766Z",
    "textMessageRequest": {
      "content": {"text": "Hello from server"}
    }
  },
  "date": "2026-03-26T16:51:36Z",
  "user_id": "c2420cb3-08eb-4ca4-b299-7c04c195f413",
  "status": "IN_DELIVERY"
}
''';

const _ackFrame = '''
{
  "type": "CONNECTION_ACK_TYPE_CONNECTION_ACK",
  "connectionId": "conn-1",
  "timestamp": "2026-03-26T16:51:30Z"
}
''';

void main() {
  group('YaloMessageServiceWebSocket', () {
    late MockYaloMessageAuthService auth;
    late List<_FakeChannel> channels;
    late YaloMessageServiceWebSocket service;

    WebSocketChannel factory(Uri uri) {
      final ch = _FakeChannel(uri);
      channels.add(ch);
      return ch;
    }

    setUp(() {
      auth = MockYaloMessageAuthService();
      channels = [];
      service = YaloMessageServiceWebSocket(
        baseUrl: 'api.example.com',
        authService: auth,
        channelFactory: factory,
      );
    });

    tearDown(() => service.dispose());

    test('opens websocket with token query parameter', () async {
      when(
        () => auth.auth(),
      ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

      service.messages().listen((_) {});
      await Future.delayed(Duration.zero);

      expect(channels, hasLength(1));
      expect(
        channels.single.uri.toString(),
        equals(
          'wss://api.example.com/websocket/v1/connect/inapp?token=abc',
        ),
      );
    });

    test('encodes the access token for URL safety', () async {
      when(
        () => auth.auth(),
      ).thenAnswer((_) async => Result.ok(_tokenEntry('a b/c?d=1')));

      service.messages().listen((_) {});
      await Future.delayed(Duration.zero);

      expect(
        channels.single.uri.queryParameters['token'],
        equals('a b/c?d=1'),
      );
    });

    test('emits decoded PollMessageItem for incoming text frames', () async {
      when(
        () => auth.auth(),
      ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

      final received = <PollMessageItem>[];
      service.messages().listen(received.add);
      await Future.delayed(Duration.zero);

      channels.single.emit(_ackFrame);
      channels.single.emit(_incomingFrame);
      await Future.delayed(Duration.zero);

      expect(received, hasLength(1));
      expect(
        received.single.message.textMessageRequest.content.text,
        equals('Hello from server'),
      );
    });

    test('drops PollMessageItem frames that arrive before the connection ack',
        () async {
      when(
        () => auth.auth(),
      ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

      final received = <PollMessageItem>[];
      service.messages().listen(received.add);
      await Future.delayed(Duration.zero);

      channels.single.emit(_incomingFrame);
      await Future.delayed(Duration.zero);

      expect(received, isEmpty);
    });

    test('ignores malformed frames without crashing', () async {
      when(
        () => auth.auth(),
      ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

      final received = <PollMessageItem>[];
      service.messages().listen(received.add);
      await Future.delayed(Duration.zero);

      channels.single.emit(_ackFrame);
      channels.single.emit('not json');
      channels.single.emit('"a string, not an object"');
      channels.single.emit(_incomingFrame);
      await Future.delayed(Duration.zero);

      expect(received, hasLength(1));
    });

    test('buffers frames until the connection ack arrives', () async {
      when(
        () => auth.auth(),
      ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

      service.messages().listen((_) {});
      await Future.delayed(Duration.zero);

      final beforeAck = await service.sendSdkMessage(_textMessage('one'));
      expect(beforeAck, isA<Ok<Unit>>());
      expect(channels.single.sent, isEmpty);

      channels.single.emit(_ackFrame);
      await Future.delayed(Duration.zero);

      final afterAck = await service.sendSdkMessage(_textMessage('two'));
      expect(afterAck, isA<Ok<Unit>>());
      expect(channels.single.sent, hasLength(2));
      final firstBody =
          jsonDecode(channels.single.sent[0] as String)
              as Map<String, dynamic>;
      final secondBody =
          jsonDecode(channels.single.sent[1] as String)
              as Map<String, dynamic>;
      expect(
        firstBody['textMessageRequest']['content']['text'],
        equals('one'),
      );
      expect(
        secondBody['textMessageRequest']['content']['text'],
        equals('two'),
      );
    });

    test('buffers frames again after a close and flushes them on the next ack',
        () {
      fakeAsync((async) {
        when(
          () => auth.auth(),
        ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

        service.messages().listen((_) {});
        async.flushMicrotasks();
        channels.single.emit(_ackFrame);
        async.flushMicrotasks();

        channels.single.closeStream();
        async.flushMicrotasks();

        service.sendSdkMessage(_textMessage('after-close'));
        async.flushMicrotasks();

        async.elapse(const Duration(seconds: 1));
        async.flushMicrotasks();
        expect(channels, hasLength(2));
        expect(channels.last.sent, isEmpty);

        channels.last.emit(_ackFrame);
        async.flushMicrotasks();

        expect(channels.last.sent, hasLength(1));
        final body =
            jsonDecode(channels.last.sent.single as String)
                as Map<String, dynamic>;
        expect(
          body['textMessageRequest']['content']['text'],
          equals('after-close'),
        );
      });
    });

    test(
      'sendSdkMessage returns Error when called before any listener subscribes',
      () async {
        when(
          () => auth.auth(),
        ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

        final result = await service.sendSdkMessage(_textMessage('hi'));
        await Future.delayed(Duration.zero);

        expect(result, isA<Error<Unit>>());
        expect(channels, isEmpty);
      },
    );

    test('reconnects with exponential backoff after the socket closes',
        () async {
      fakeAsync((async) {
        when(
          () => auth.auth(),
        ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

        service.messages().listen((_) {});
        async.flushMicrotasks();
        expect(channels, hasLength(1));

        channels.first.closeStream();
        async.flushMicrotasks();

        async.elapse(const Duration(milliseconds: 999));
        expect(channels, hasLength(1));

        async.elapse(const Duration(milliseconds: 2));
        async.flushMicrotasks();
        expect(channels, hasLength(2));

        channels.last.closeStream();
        async.flushMicrotasks();
        async.elapse(const Duration(seconds: 2));
        async.flushMicrotasks();
        expect(channels, hasLength(3));
      });
    });

    test('schedules a reconnect when auth fails', () {
      fakeAsync((async) {
        when(
          () => auth.auth(),
        ).thenAnswer((_) async => Result.error(Exception('unauthorized')));

        service.messages().listen((_) {});
        async.flushMicrotasks();
        expect(channels, isEmpty);

        async.elapse(const Duration(seconds: 1));
        async.flushMicrotasks();
        expect(verify(() => auth.auth()).callCount, greaterThanOrEqualTo(2));
      });
    });

    test('schedules a reconnect when the stream errors', () {
      fakeAsync((async) {
        when(
          () => auth.auth(),
        ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

        service.messages().listen((_) {});
        async.flushMicrotasks();
        expect(channels, hasLength(1));

        channels.first.emitError(Exception('boom'));
        async.flushMicrotasks();
        async.elapse(const Duration(seconds: 1));
        async.flushMicrotasks();

        expect(channels, hasLength(2));
      });
    });

    test('dispose stops reconnects and closes the active socket', () {
      fakeAsync((async) {
        when(
          () => auth.auth(),
        ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

        service.messages().listen((_) {});
        async.flushMicrotasks();
        final firstChannel = channels.first;

        service.dispose();
        async.elapse(const Duration(seconds: 5));
        async.flushMicrotasks();

        expect(firstChannel.sinkClosed, isTrue);
        expect(channels, hasLength(1));
      });
    });

    test('pause closes the socket without reconnecting', () {
      fakeAsync((async) {
        when(
          () => auth.auth(),
        ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

        service.messages().listen((_) {});
        async.flushMicrotasks();
        final firstChannel = channels.first;

        service.pause();
        async.elapse(const Duration(seconds: 5));
        async.flushMicrotasks();

        expect(firstChannel.sinkClosed, isTrue);
        expect(channels, hasLength(1));
      });
    });

    test('resume reconnects after a pause', () {
      fakeAsync((async) {
        when(
          () => auth.auth(),
        ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

        service.messages().listen((_) {});
        async.flushMicrotasks();
        expect(channels, hasLength(1));

        service.pause();
        service.resume();
        async.flushMicrotasks();

        expect(channels, hasLength(2));
      });
    });

    test('resume is a no-op when not paused', () {
      fakeAsync((async) {
        when(
          () => auth.auth(),
        ).thenAnswer((_) async => Result.ok(_tokenEntry('abc')));

        service.messages().listen((_) {});
        async.flushMicrotasks();

        service.resume();
        async.flushMicrotasks();

        expect(channels, hasLength(1));
      });
    });
  });
}
